### ISSUE TITLE
Should the Messages page automatically unhighlight messages?

### ISSUE THREAD

**Issue by towerofnix** [LLK/scratch-www#1547]
(TL;DR, the messages page doesn't highlight the right messages. Ideas for a solution are given.)

Currently a trouble I'm having with the migrated Messages page is that when it refreshes the message count, it *un-*highlights any of the messages that were highlighted when I first opened the page. This was a problem because I had forgotten to check my new notifications; when I went back to my Messages tab, the notification list displayed as if there were no new messages. I had to remember that the notification bubble said 4; if I hadn't, I might not have thought I had any new notifications (and so I'd have missed those 4 messages).

---

A more likely case:

I see that I have four new notifications by looking at the navigation bar, so I go to my Messages page. All four notifications are "there are new posts in the forum thread" messages, so I open one of the forum thread links in a separate tab and read the new posts. This takes roughly five minutes. Now that I'm done reading the new forum posts, I close that tab. I see my Messages tab again.. and no messages are highlighted! This is a bit of a nuisance, since I have to attempt to remember how many notifications I had when loading the page (Three? Four? Or only the one?). To make things worse, I might close the Messages page thinking I *already viewed* all of the notifications, when I only read one or two; that means I'd miss some!

---

I'm not certain what the best solution is here, but the current behavior is *definitely* wrong:

* When the Messages page is opened, fetch a list of messages.
* When the Messages page is opened, fetch the current message count.
* Display all messages, highlighting the top (current message count) messages.
* Clear the user's message count (reset it to 0, via the API).
* Set an interval to get the number of new messages every 120 seconds. When this interval runs, highlight the top (current message) count. Note that this DOES NOT reload the actual list of messages.

This is wrong because:

* If the user's message count is zero when the interval runs (which is likely, because the notification count is cleared after it is gotten when loading the Messages page), the new messages are no longer highlighted. Both the situations I described above are possible.

* If the user's message count is greater than zero less than the message count when the Messages page was loaded, the user may forget they got one of the notifications. (For example, if I had 4 notifications, then loaded my Messages page, then forgot to read it, and somebody loved 3 of my projects, then the first three messages from the list gotten when I loaded the Messages page would be highlighted, and I'd miss the fourth notification!)

* If the user's message count when the interval runs is *greater than* the message count when the Messages page was loaded, then an *old* notification would end up being highlighted. It would likely be one the user had already read earlier; and this would inevitably lead to confusion.

---

Some ideas for a solution:

* Don't run the message count refresh interval on the Messages page. This is an easy option, and would get rid of all the above problems.

* Getting rid of the refresh interval on the page altogether wouldn't be entirely good, though; the user wouldn't be able to tell how many new notifications they have from the Messages page. Instead, it might be better to simply not re-highlight the list based on the new message count. It would also be important to not update the count in the "bubble" next to the "Messages" list heading. The new message count would still be displayed next to the mail icon in the navbar, so the user could still quickly see how many new notifications they have.

* Plus, disabling the message count refresh interval altogether on the Messages page would make the navigation bar behave differently based on the page you're on. This doesn't make logical sense; why should a global element behave differently based on the active page? It's also confusing to users. Why is my message count be different on my Messages page?

* Adding onto the previous point, I believe it would make the most sense if the notification count in the navbar was reset to 0 (i.e. the bubble hidden) as soon as the Messages page is loaded. This makes sense because the navbar is used to show the number of *new* messages you have; when you load the Messages page, that number is reset to 0 via the API - so, on new pages you open, it will be 0 - but the navbar on the Messages page is not updated to reflect this until 120 seconds later (when the message count refresh interval runs). [A Scratch user was confused by this.](https://scratch.mit.edu/discuss/topic/276006/)

* But this whole change could be confusing to users in its own way. Disconnecting the messages bubble in the navbar and the bubble next to the "Messages" list heading is likely going to be surprising, since they look so similar. But there are some ways to make it less confusing:

  * If the navbar message count is immediately reset to 0 when loading the Messages page, it would be immediately obvious that the counter next to "Messages" isn't the same as the one in the navbar.

  * By displaying the text "(# new - refresh the page)" (or similar) next to the bubble beside "Messages", it would be obvious that new messages could be read as soon as the page is refreshed. (This text would, of course, act as a link to refresh the notification list.)

---

Sorry for the rather long issue! But I think these details and ideas could be helpful in finding the right solution.

**Comment by mewtaylor** [comment]
@towerofnix please don't apologize for the length of this issue – it's so thorough, I really appreciate it!

Of the solutions you proposed, I think the one I like the best is:
* When the interval is run, compare the new count to the previous count
  * If it's now 0 and was >0 before, set the count to 0 and stop highlighting unread messages
  * If it's now a different number >0 and was >0 before, mark messages as read, and display a message along the lines of ("Please refresh this page to see new notifications"). Have the bubble in the nav display the new notification count
  * If it was at 0 and is now >0, display a message along the lines of ("Please refresh this page to see new notifications"). Have the bubble in the nav display the new notification count

@carljbowman do you have any thoughts on this UX?

@thisandagain @rschamp I'd propose moving this to October so we can include this with other notifications page UX enhancements discussed in other issues, if that's okay.

**Comment by towerofnix** [comment]
> If it's now 0 and was >0 before, set the count to 0 and stop highlighting unread messages.

I think this would still cause the issue cases I described? - Particularly "and stop highlighting unread messages", since, when you go back to check the Messages tab after spending +120 seconds away from it (e.g. because you were reading another message, or had to temporarily leave the computer), the messages would be unhighlighted.

**Comment by mewtaylor** [comment]
@towerofnix ah I may have misread the issues you mentioned – which issue you described would it cause? Or is this the issue you were describing with not being able to remember that a notification was read if you leave the page to look at a notification and come back to it?

**Comment by towerofnix** [comment]
>Or is this the issue you were describing with not being able to remember that a notification was read if you leave the page to look at a notification and come back to it?

@mewtaylor Yeah, that's the one I was talking about. (First two paragraphs "Currently a trouble..." and "I see that I have four new notifications..." are example cases.)
